faculty/clients/environment.py

```diff
@@ -30,9 +57,126 @@
         "author_id",
         "created_at",
         "updated_at",
+        "specification",
     ],
 )
 
+EnvironmentUpdate = namedtuple(
+    "EnvironmentUpdate", ["name", "description", "specification"]
+)
+
+
+class SemverSchema(BaseSchema):
+    constraint = EnumField(Constraint, by_value=True, required=True)
+    identifier = fields.String(required=True)
+
+    @post_load
+    def make_version(self, data):
+        return Version(**data)
+
+
+class VersionField(fields.Field):
+    """
+    Field that serialises/deserialises a Python package version.
+    """
+
+    def _deserialize(self, value, attr, obj, **kwargs):
+        if value == "latest":
+            return "latest"
+        else:
+            return SemverSchema().load(value)
+
+    def _serialize(self, value, attr, obj, **kwargs):
+        if value == "latest":
+            return "latest"
+        else:
+            return SemverSchema().dump(value)
+
+
+class PythonPackageSchema(BaseSchema):
+    name = fields.String(required=True)
+    version = VersionField(required=True)
+
+    @post_load
+    def make_python_package(self, data):
+        return PythonPackage(**data)
+
+
+class PipSchema(BaseSchema):
+    extra_index_urls = fields.List(
+        fields.String(), data_key="extraIndexUrls", required=True, many=True
+    )
+    packages = fields.List(fields.Nested(PythonPackageSchema()), required=True)
+
+    @post_load
+    def make_pip(self, data):
+        return Pip(**data)
+
+
+class CondaSchema(BaseSchema):
+    channels = fields.List(fields.String(), required=True)
+    packages = fields.List(fields.Nested(PythonPackageSchema()), required=True)
+
+    @post_load
+    def make_conda(self, data):
+        return Conda(**data)
+
+
+class PythonSpecificationSchema(BaseSchema):
+    conda = fields.Nested(CondaSchema(), required=True)
+    pip = fields.Nested(PipSchema(), required=True)
+
+    @post_load
+    def make_python_specification(self, data):
+        return PythonSpecification(**data)
+
+
+class PythonSchema(BaseSchema):
```

Comment:
There already is a `PythonSpecificationSchema` (used to deserialise the `Python2` and `Python3` keys). Not entirely happy with the current naming of `PythonSchema`, but not sure what a good alternative should be.

Reply:
In reality, the Python2 and Python3 keys are not exclusive - Python2 and Python3 are just the names of the default environment names - you can also specify others.

This may imply changing the current PythonSpecficationSchema to PythonEnvironmentSchema and/or removing the current PythonSchema and replacing it in SpecificationSchema to a `fields.Dict(fields.String(), fields.Nested(PythonSpecificationSchema()))`? Happy to discuss this in person to save time.

Reply:
Ok, let's leave the structure as is, but my final suggestion would be to do the following renames:

- `PythonSpecificationSchema` -> `PythonEnvironmentSchema`
- `PythonSchema` -> `PythonSpecificationSchema`

As this communicates the intended meaning of the former to represent a Python conda environment.